gift-certificate-data/src/main/java/com/epam/esm/giftcertificatemodule3/dao/impl/GiftCertificateDAOHibernate.java

```diff
@@ -2,54 +2,142 @@
+
+        final List<Predicate> andPredicates = new ArrayList<>();
+
+        if (searchParametersHolder.getTagName() != null) {
+            andPredicates.add(cb.equal(tagRoot.get("name"), searchParametersHolder.getTagName()));
+        }
+        if (searchParametersHolder.getName() != null) {
+            andPredicates.add(cb.like(tagGiftCertificateJoin.get("name"),
+                    "%" + searchParametersHolder.getName() + "%"));
+        }
+        if (searchParametersHolder.getDescription() != null) {
+            andPredicates.add(cb.like(tagGiftCertificateJoin.get("description"),
+                    "%" + searchParametersHolder.getDescription() + "%"));
+        }
+        if (searchParametersHolder.getSortBy() != null && sortDesc) {
+            cq.orderBy(cb.desc(tagGiftCertificateJoin.get(searchParametersHolder.getSortBy())));
+        } else if (searchParametersHolder.getSortBy() != null) {
+            cq.orderBy(cb.asc(tagGiftCertificateJoin.get(searchParametersHolder.getSortBy())));
+        }
+
+        cq.select(tagGiftCertificateJoin)
+                .where(andPredicates.toArray(new Predicate[andPredicates.size()]))
+                .distinct(true);
+
+        TypedQuery<GiftCertificate> query = em.createQuery(cq);
+        query.setFirstResult(firstResult);
+        query.setMaxResults(maxResults);
+        List<GiftCertificate> certificates = query.getResultList();
+
+        return certificates;
+    }
+
+    @Override
+    public void update(GiftCertificate giftCertificate) {
+        EntityManager em = emf.createEntityManager();
+        em.joinTransaction();
+        GiftCertificate oldGiftCertificate = findById(giftCertificate.getId());
+        setUpdatedFields(giftCertificate, oldGiftCertificate);
     }
 
     @Override
     public void delete(GiftCertificate object) {
-        Session session = entityManager.unwrap(Session.class);
-        session.delete(object);
+        EntityManager em = emf.createEntityManager();
+        em.joinTransaction();
+        em.remove(object);
     }
 
     @Override
     public void deleteById(Long id) {
-        Session session = entityManager.unwrap(Session.class);
-        Query query = session.createQuery("delete from GiftCertificate where id_cert=:id");
-        query.setParameter("id_cert", id);
+        EntityManager em = emf.createEntityManager();
+        em.joinTransaction();
+        Query query = em.createQuery("delete from GiftCertificate c where c.id=:id");
+        query.setParameter("id", id);
         query.executeUpdate();
     }
+
+    private void setUpdatedFields(GiftCertificate giftCertificate, GiftCertificate oldGiftCertificate) {
+        if (giftCertificate.getTags() == null) {
+            giftCertificate.setTags(oldGiftCertificate.getTags());
+        }
+        if (giftCertificate.getName() == null) {
+            giftCertificate.setName(oldGiftCertificate.getName());
+        }
+        if (giftCertificate.getDescription() == null) {
```

Comment:
Just a nitpick. Description is nullable column. The user will not be able to update description to be null.